### DISCUSSION TITLE
After upgrading from v8 to v9, I get the error "Error: size: 1" when using Tile3DLayer in MapboxOverlay

### DISCUSSION THREAD

**Discussion by hig-bit** [visgl/deck.gl#8815] [Q&A]
An error will occur with the 3D tiles downloaded from the following page.
[https://www.geospatial.jp/ckan/dataset/plateau-13103-minato-ku-2023/resource/5f661701-f928-431a-9120-1ab46045bbf5](https://www.geospatial.jp/ckan/dataset/plateau-13103-minato-ku-2023/resource/5f661701-f928-431a-9120-1ab46045bbf5)

No error will occur with the 3D tiles downloaded from the following page. It will be drawn properly.
[https://github.com/CesiumGS/3d-tiles-samples/tree/main/1.0/TilesetWithDiscreteLOD](https://github.com/CesiumGS/3d-tiles-samples/tree/main/1.0/TilesetWithDiscreteLOD)

The contents of the error are as follows.

```
deck: initialization of ScenegraphLayer({id: 'plateau123-scenegraph-http://127.0.0.1:3000/minato-ku/lod2/data/data685.b3dm'}): size: 1 Error: size: 1
    at $0 (dist.min.js:106:14179)
    at YV (dist.min.js:1676:30653)
    at GC (dist.min.js:1676:30083)
    at Wx.setGeometry (dist.min.js:1676:38847)
    at new Wx (dist.min.js:1676:36604)
    at o2 (dist.min.js:3083:534)
    at Ig.createPrimitive (dist.min.js:3084:1173)
    at dist.min.js:3084:935
    at Array.map (<anonymous>)
    at Ig.createMesh (dist.min.js:3084:919)
```

Source code

```html
<!DOCTYPE html>
<html lang="ja">
<head>
<title>deck.gl</title>
<link rel="icon" href="#" />
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<link href="https://api.mapbox.com/mapbox-gl-js/v3.3.0/mapbox-gl.css" rel="stylesheet">
<script src="https://api.mapbox.com/mapbox-gl-js/v3.3.0/mapbox-gl.js"></script>
<script src="https://unpkg.com/deck.gl@8.9.35/dist/dist.dev.js"></script>
<!-- <script src="https://unpkg.com/deck.gl@9.0.8/dist.min.js"></script> -->
<script src="https://cdn.tailwindcss.com"></script>
</head>
<body>
<div id="map" class="w-full h-screen"></div>
<script>
const { Tile3DLayer, MapboxOverlay, LightingEffect, DirectionalLight, AmbientLight } = deck;
const map = new mapboxgl.Map({
  accessToken: "<Mapbox Access Token>",
  container: "map",
  style: "mapbox://styles/mapbox/outdoors-v12",
  center: [139.741, 35.646], // minato-ku
  // center: [-75.61, 40.04], // 3d-tiles-samples
  zoom: 14,
  pitch: 60,
  bearing: 0,
});
const deckOverlay = new MapboxOverlay({
  interleaved: false,
  effects: [
    new  LightingEffect({
      ambientLight: new AmbientLight({
        color: [255, 255, 255],
        intensity: 2,
      }),
      directionalLight: new DirectionalLight({
        color: [255, 255, 255],
        intensity: 2,
        direction: [ 100, 100, 100]
      }),
    }),
  ],
  layers: [
    new Tile3DLayer({
      id: 'plateau123',
      data: "/minato-ku/lod2/tileset.json", // Error occurs when using v9
      // data: './3d-tiles-samples/1.0/TilesetWithDiscreteLOD/tileset.json', // no error occurs
    }),
  ]
});
map.on("load", async () => {
  map.addControl(deckOverlay);
});
</script>
</body>
</html>
```

It may be a problem with the 3D Tiles file format, but is there a way to draw correctly even in v9?

**Comment by hig-bit** [comment]
The 3DTiles in error were created from CityGML using [FME](https://fme.safe.com/).
[Japanese] https://github.com/Project-PLATEAU/FMEscript-CityGML-to-3DTiles

**Comment by ibgreen** [comment]
@belom88 It may be that 3D tiles depends to literally on luma.gl v8 accessors. Any chance your team could do a quick test on the tileset in question?

**Reply by hig-bit** [reply]
Thank you.

**Comment by hig-bit** [comment]
Failure and normal examples have been uploaded.

Version 9.0.14
[https://deck-gl-v9-plateau.pages.dev/](https://deck-gl-v9-plateau.pages.dev/)
<img width="1912" alt="image" src="https://github.com/visgl/deck.gl/assets/82851702/170fbe51-076f-48d8-aeb2-4109d535791f">

3DTitles assets
- [/minato-ku/lod1/tileset.json](https://deck-gl-v9-plateau.pages.dev/minato-ku/lod1/tileset.json)
- [/minato-ku/lod1/data/data8.b3dm](https://deck-gl-v9-plateau.pages.dev/minato-ku/lod1/data/data8.b3dm)
- [/minato-ku/lod1/data/data12.b3dm](https://deck-gl-v9-plateau.pages.dev/minato-ku/lod1/data/data12.b3dm)
- [/minato-ku/lod1/data/data28.b3dm](https://deck-gl-v9-plateau.pages.dev/minato-ku/lod1/data/data28.b3dm)
- [/minato-ku/lod1/data/data29.b3dm](https://deck-gl-v9-plateau.pages.dev/minato-ku/lod1/data/data29.b3dm)
- [/minato-ku/lod1/data/data207.b3dm](https://deck-gl-v9-plateau.pages.dev/minato-ku/lod1/data/data207.b3dm)

Error messages

```
index-Dd_tDSNk.js:3273 deck: initialization of ScenegraphLayer({id: 'plateau123-scenegraph-https://deck-gl-v9-plateau.pages.dev/minato-ku/lod1/data/data29.b3dm'}): size: 1 Error: size: 1
    at GV (index-Dd_tDSNk.js:2209:14869)
    at HH (index-Dd_tDSNk.js:3266:28900)
    at VH (index-Dd_tDSNk.js:3266:28321)
    at _0.setGeometry (index-Dd_tDSNk.js:3266:38087)
    at new _0 (index-Dd_tDSNk.js:3266:35702)
    at Yq (index-Dd_tDSNk.js:3524:539)
    at Jq.createPrimitive (index-Dd_tDSNk.js:3525:1218)
    at index-Dd_tDSNk.js:3525:978
    at Array.map (<anonymous>)
    at Jq.createMesh (index-Dd_tDSNk.js:3525:962)
```

Version 8.9.35
[https://deck-gl-v9-plateau.pages.dev/](https://deck-gl-v8-plateau.pages.dev/)
<img width="1912" alt="image" src="https://github.com/visgl/deck.gl/assets/82851702/ceaf5713-58c0-45f8-a596-d46a1d389a6e">

**Reply by hig-bit** [reply]
Source code is here.
[high-u/deckgl-plateau](https://github.com/high-u/deckgl-plateau)

**Comment by eoudejans** [comment]
I encounter a very similar error, it seems related to incorrect byte alignment for uint8 (in my case) vertex attribute not aligned to 16bit boundaries see [WebGPU data memory layout](https://webgpufundamentals.org/webgpu/lessons/webgpu-memory-layout.html):

See code in @luma.gl\core\dist\adapter\type-utils\vertex-format-from-attribute.js line 62:
```ts
/** Get the vertex format for an attribute with TypedArray and size */
export function getVertexFormatFromAttribute(typedArray, size, normalized) {
    if (!size || size > 4) {
        throw new Error(`size ${size}`);
    }
    const components = size;
    let dataType = getDataTypeFromTypedArray(typedArray);
    if (dataType === 'uint8' || dataType === 'sint8') {
        if (components === 1 || components === 3) {
            // WebGPU 8 bit formats must be aligned to 16 bit boundaries');
            throw new Error(`size: ${size}`);
        }
        if (normalized) {
            dataType = dataType.replace('int', 'norm');
        }
        return `${dataType}x${components}`;
    }
    
    ...
```

**Comment by eoudejans** [comment]
Following the luma.gl v8 to v9 breaking change guide it shows that the WebGL2 API (and support) is exchanged for WebGPU, 
https://luma.gl/docs/legacy/porting-guide#why-is-the-v9-api-a-major-breaking-change. The 16bit alignment may be a consequence of WebGPU strictness compared to WebGL2 API.

**Comment by eoudejans** [comment]
Load option: 3d-tiles.decodeQuantizedPositions in https://loaders.gl/docs/modules/3d-tiles/api-reference/tiles-3d-loader seemed to be what I needed, however no succes:

```ts
    this.tile3DLayer = new Tile3DLayer({
      id: 'tile-3d-layer',
      data: url,
      pickable: true,
      loadOptions: {
        '3d-tiles': { decodeQuantizedPositions: true }
      },
      ...
```

**Comment by eoudejans** [comment]
Following the backend notes for luma.gl:
https://luma.gl/docs/api-guide/gpu/gpu-attributes/#backend-notes

It states that:
WebGPU 8 bit integers must be aligned to 16 bits (uint8x1, uint8x3, unorm8x1, unorm8x3 etc` are not supported)

And that:
8 and 16 bit values only support 2 or 4 components. This is a WebGPU specific limitation that does not exist on WebGL.

**Reply by ibgreen** [reply]
Correct. The part that is perhaps missing is that the luma v9 API is a "WebGPU first design" that is optimized for WebGPU with added portability to WebGL. This means that it intentionally removes explicit support for options that are WebGL only, to ensure that apps remain portable.

I can't say for sure that this is the cause of the failure, but once I get time to look into this, this limitation around 1 byte attribute support could be one of the first "clues" to what is going on.

**Comment by eoudejans** [comment]
@ibgreen Thanks for your response and clarification!
Do I understand the implementation correctly in that the enforced number of components in getVertexFormatFromAttribute for uint8 and sint8 value types is solely a WebGPU constrait that could be relaxed for WebGL usage?

**Reply by ibgreen** [reply]
We could have chosen to design luma.gl such that it would provide the union of all options offered by WebGL and WebGPU and leave it to application writer to add checks all over their code and implement multiple paths.

We did not choose this approach. Instead we focused on the capabilities of WebGPU (which we consider the future) and we provide best-effort WebGL bindings to match that API.

There is usually a good reason why WebGPU is more restrictive. It turns out that a number of things that seem to be supported in WebGL are handled by code in the driver silently copying buffers in formats supported by the underlying GPU, hiding performance inefficiencies.

That said, I have heard some noise that WebGPU team might have found a way to relax this particular restriction.

**Reply by eoudejans** [reply]
Thanks a lot for this clarification, I can totally follow the chosen approach in light of the performance penalties associated with silent copying of buffers. I'm wondering how the WebGPU team are contemplating a way to relax this particular restriction.

**Reply by ibgreen** [reply]
Seem to recall seeing some notes around this, but can't find a link right now, possibly I understood incorrectly. But I could imagine it being proposed if it was determined that all target systems and underlying APIs can in fact now support this without perf penalties. 

FWIW, WebGPU is still evolving with every Chrome release https://developer.chrome.com/blog/new-in-webgpu-120#chrome127. New features, extensions, clarifications, etc.

**Reply by eoudejans** [reply]
Interesting, I will keep an eye for new developments, thanks! 

On another note, I hope that our issues have enough common ground and _added to_ instead of _hijacked_ your discussion @hig-bit.

**Reply by ibgreen** [reply]
Apologies, I suppose my chit-chat is just a way of stalling for time until someone is actually able to look at this. Figured it would be better than radio silence...

**Reply by mikoding9** [reply]
Hi there,

Is there a workaround or solution for this issue, or is downgrading to version 8 the only option?

Thank you!

**Reply by ibgreen** [reply]
No know workaround other than the observation that not all models are affected - i.e. someone might be able to control how models are generated.

I've tried to relax the restriction in the function listed above, let's see if it is an easy fix. 

Fix published in luma.gl@9.0.25

**Reply by eoudejans** [reply]
Thanks for the restriction relaxation! The variable "normalized" is undefined in my scenario unfortunately:
![image](https://github.com/user-attachments/assets/0d0a331f-cbe2-4ebe-918e-45cfd9bf357d)
 
I do think that this relaxation is an improvement.

**Comment by MariDani** [comment]
Hello everyone,
I was just wondering if there have been any updates on 3Dtiles visualisations since the last post in August 2024. Or if I need to start digging into the luma.gl WebGPU attributes.
Thank you, 
Mariana
